### rds files 

`SwitchList.rds` - pre switching, all isoforms that pass IsoformSwitchAnalyzeR's filters. 

`SwitchList_part2.rds` - switchlist with all switching consequences analyzed.  

### plots 
`Splicing_Summary.pdf` types of switching events and frequencies. 

`Splicing_Enrichment.pdf` - out of the upregulated isoforms associated to the consequence for condition 2(RGC), what proportion had consequence 1. 

  `ES`: Exon Skipping. Compared to the hypothetical pre-RNA a single exon was skipped in the isoform analyzed (for every ES event annotated).
  `MEE`: Mutually exclusive exon. Special case were two isoforms form the same gene contains two mutually exclusive exons and which are not found in any of the other isoforms from that gene.
  `MES`: Multiple Exon Skipping. Compared to the hypothetical pre-RNA multiple consecutive exon was skipped in the isoform analyzed (for every MES event annotated).
  `IR`: Intron Retention. Compared to the hypothetical pre-RNA an intron was retained in the isoform analyzed.
  `A5`: Alternative 5'end donor site. Compared to the hypothetical pre-RNA an alternative 5'end donor site was used. Since it is compared to the pre-RNA, the donor site used is per definition more upstream than the the pre-RNA (the upstream exon is shorter).
  `A3`: Alternative 3'end acceptor site. Compared to the hypothetical pre-RNA an alternative 3'end acceptor site was used. Since it is compared to the pre-RNA, the donor site used is per definition more downstream than the the pre-RNA (the downstream exon is shorter).
  `ATSS`: Alternative Transcription Start Sites. Compared to the hypothetical pre-RNA an alternative transcription start sites was used. Since it is compared to the pre-RNA, the ATSS site used is per definition more downstream than the the pre-RNA .
  `ATTS`: Alternative Transcription Termination Sites. Compared to the hypothetical pre-RNA an alternative transcription Termination sites was used. Since it is compared to the pre-RNA, the ATTS site used is per definition more upstream than the the pre-RNA.

`common_switch_consequences.pdf` - common switching consequences, including domain loss, gain, IR, signal peptide, etc.

`DGE_DTU_DTE_venn.pdf` - venn diagram of genes 
with differential gene expression, differential transcript usage, and differential transcript expression events.

`DGE_DTU_DTE_barplot.pdf` - proportional barplot of DTU with DGE/non-DGE and DTE/non-DTE genes.


In `part2` folder: switch plots of top 50 DTU genes with switch consequences.

### FASTA files 
`AA fasta` files - generated by `IsoformSwitchAnalyzeR`, containing protein sequences of isoforms, used by `SignalP` and `Pfam`.

`nt fasta` files - generated by `IsoformSwitchAnalyzeR`, containing nucleotide sequences of isoforms.

### outputs from Pfam, SignalP and CPC2
`pfam_results.txt`, `pfam_results.csv`
`region_output.gff3`, `processed_entries.fasta`, 
`prediction_results.txt`, `output.json`, `output.gff`, `CPC2_output.txt`

### tsv files 
`DTE_table.tsv` - table of all DTE events.
`DGE_table.tsv` - table of all DGE events.
`DGE_DTU_DTE.tsv` - table of all DGE, DTU, and DTE events.
`retnet_dtu_genes.tsv` - DTU genes that were found in RetNet.






