---
---
---

## FASTQ processing

### Analysis 1: `code/01_fastq_processing/02_MinIONQC.sh`

-   **Question**: Read statistics - read length, base quality, N50 distribution
-   **Input**: Raw FASTQ
-   **Method**: MinIONQC was ran.
-   **Output**: summary yaml and individual plots on read length distribution, base quality distribution, yield over time

### Analysis 2: `code/01_fastq_processing/03_fastq_qc.sh`

-   **Question**: Remove low quality reads based on Phred score.
-   **Input**: Raw FASTQ, counts matrix, etc.
-   **Method**: Nanofilt removes reads in fastq based on ONT summary text file.
-   **Output**: processed FASTQ

### Analysis 3: `code/01_fastq_processing/minionQC_yaml.R`

-   **Question**: Produce boxplots of mean read length, median q value, median N50 and total number of reads across all samples.
-   **Input**: YAML summary file produced by MinIONQC.
-   **Method**: ggplot2
-   **Output**: boxplots

## FASTQ to BAM

### Analysis 4: `code/01b_fastq_to_bam/01_fastq_to_bam_gencode_splice.sh`

-   **Question**: Alignment with genome
-   **Input**: Nanofilt processed FASTQ
-   **Method**: Minimap2 was used to align reads to the genome.
-   **Output**: bam

### Analysis 5: `code/01b_fastq_to_bam/01b_fastq_to_bam_transcriptome_gencode.sh`

-   **Question**: Alignment with transcriptome
-   **Input**: Nanofilt processed FASTQ
-   **Method**: Minimap2 was used to align reads to the transcriptome.
-   **Output**: bam

### Analysis 6: `code/01b_fastq_to_bam/02_high_quality_bam_genome.sh`

-   **Question**: Removes alignments with MAPQ \< 30. Only keeps primary mapped alignments in chr 1-22, X, Y, M.
-   **Input**: Nanofilt processed FASTQ
-   **Method**: samtools was used to filter the bam file, and create flagstat summary.
-   **Output**: bam, summary stats on alignments.

## BAM QC Visualization

### Analysis 7: `code/01b_fastq_to_bam/02_high_quality_bam_genome.sh`

-   **Question**: Removes alignments with MAPQ \< 30. Only keeps primary mapped alignments in chr 1-22, X, Y, M.
-   **Input**: Nanofilt processed FASTQ
-   **Method**: samtools was used to filter the bam file, and create flagstat summary.
-   **Output**: bam, summary stats on alignments.

### Analysis 8: `code/02_bam_QC/01a_multi_exon_pcg_sample_specific.sh`

-   **Question**: Exon-exon junction distribution
-   **Input**: bam files produced in analysis 5
-   **Method**: Multi-exon (PCG and all genes) junctions were quantified using python script `01_multi_exon_pcg.py` and `02_multi_exon_all_genes.py`. Visualization was done using R script `01_multi_exon_pcg.R` and `code/02_bam_QC/exon_exon_boxplots.R`
-   **Output**: plots

### Analysis 9: `code/02_bam_QC/read_type_percentages.R`

-   **Question**: Percentage of alignments that are primary, supplementary, and unmapped.
-   **Input**: `flagstat` file produced in analysis 4: `/retina_lrs/05_bams/genome/primary_assembly/logs/all.flagstat.txt`
-   **Method**: R script `read_type_percentages.R` was used to parse the flagstat file and create a bar plot of read types.
-   **Output**: plots

### Analysis 10:

-   **Question**: Percentage of alignments that are primary, supplementary, and unmapped.
-   **Input**: `flagstat` file produced in analysis 4: `/retina_lrs/05_bams/genome/primary_assembly/logs/all.flagstat.txt`
-   **Method**: R script `read_type_percentages.R` was used to parse the flagstat file and create a bar plot of read types.
-   **Output**: plots

### Quantification with `code/03_quantification/05_bambu`

## Analysis 11: `code/03_quantification/05_bambu/01_generate_sample_wise_read_class.sh`

-   **Question**: Read class RDS files are generated for each sample individually.
-   **Input**: `bam` files from analysis 6.
-   **Method**: R script `read_type_percentages.R` was used to parse the flagstat file and create a bar plot of read types.
-   **Output**: `rds` files

## Analysis 12: `code/03_quantification/05_bambu/02_bambu_generate_rcs.sh`

-   **Question**: Read classes are analyzed together for all samples to produce a common extended notation.
-   **Input**: `rds` files from analysis 11.
-   **Method**: R script `read_type_percentages.R` was used to parse the flagstat file and create a bar plot of read types.
-   **Output**: final `se` object, counts matrix, extended annotation gtf.

## Analysis 13: `code/03_quantification/05_bambu/03_bambu_quantification.R`

-   **Question**: Read classes are analyzed together for all samples to produce a common extended notation.
-   **Input**: `rds` files from analysis 11.
-   **Method**: R script `read_type_percentages.R` was used to parse the flagstat file and create a bar plot of read types.
-   **Output**: intermediate `se` object

`03_bambu_quantification.R, 04_switch_plots.R`, should be archived.

## Analysis 14: `code/03_quantification/05_bambu/03_bambu_quantification.R`

-   **Question**: Read classes are analyzed together for all samples to produce a common extended notation.
-   **Input**: `rds` files from analysis 11.
-   **Method**: R script `read_type_percentages.R` was used to parse the flagstat file and create a bar plot of read types.
-   **Output**: intermediate `se` object
